File: packages/telescope-server/src/lsp/handlers/symbols.ts
/**
 * Symbols Handler
 *
 * Provides document and workspace symbols for OpenAPI documents by combining:
 * 1. YAML language service symbols (fallback - basic structure)
 * 2. OpenAPI-specific symbols (preferred - semantic operations, schemas, etc.)
 *
 * @module lsp/handlers/symbols
 */

import type { Connection, TextDocuments } from "vscode-languageserver";
import type { TextDocument } from "vscode-languageserver-textdocument";
import type {
	DocumentSymbol,
	WorkspaceSymbol,
	Range,
} from "vscode-languageserver-protocol";
import { SymbolKind } from "vscode-languageserver-protocol";

import { joinPointer } from "../../engine/utils/pointer-utils.js";
import type { DocumentCache, CachedDocument } from "../document-cache.js";
import type { TelescopeContext } from "../context.js";
import { createDocumentProvider } from "../services/document-provider.js";
import type { DocumentProvider, ProvidedDocument } from "../services/document-provider.js";
import { getYAMLService } from "../services/yaml-service.js";
import type { WorkspaceProject } from "../workspace/workspace-project.js";
import { isOpenAPIDocument } from "./shared.js";

/**
 * Register symbol handlers on the connection.
 */
export function registerSymbolHandlers(
	connection: Connection,
	documents: TextDocuments<TextDocument>,
	cache: DocumentCache,
	ctx: TelescopeContext,
	getProject: () => WorkspaceProject,
): void {
	const logger = ctx.getLogger("Symbols");
	const yamlService = getYAMLService(logger);

	// Document symbols
	connection.onDocumentSymbol((params): DocumentSymbol[] => {
		try {
			const doc = documents.get(params.textDocument.uri);
			if (!doc) return [];

			// 1. Try OpenAPI symbols first (more semantic)
			const cached = cache.get(doc);
			if (isOpenAPIDocument(cached)) {
				const openapiSymbols = provideOpenAPIDocumentSymbols(cached, cache, ctx);
				if (openapiSymbols.length > 0) {
					return openapiSymbols;
				}
			}

			// 2. Fall back to YAML service symbols
			const yamlSymbols = yamlService.getDocumentSymbols(doc);
			return yamlSymbols;
		} catch (error) {
			logger.error(
				`Document symbols failed: ${error instanceof Error ? error.message : String(error)}`,
			);
			return [];
		}
	});

	// Workspace symbols with partial result support
	connection.onWorkspaceSymbol(async (params): Promise<WorkspaceSymbol[]> => {
		try {
			const provider = createDocumentProvider({
				documents,
				cache,
				project: getProject(),
			});

			// Create a partial result reporter if the client supports it
			const partialResultToken = params.partialResultToken;
			let partialReporter: ((symbols: WorkspaceSymbol[]) => void) | undefined;

			if (partialResultToken) {
				partialReporter = (symbols: WorkspaceSymbol[]) => {
					// Send partial results to the client
					connection.sendProgress(
						// Use the native type - any string/number token is valid
						{ method: "workspace/symbol" } as any,
						partialResultToken,
						symbols,
					);
				};
			}

			return await provideWorkspaceSymbols(
				params.query,
				ctx,
				provider,
				partialReporter,
			);
		} catch (error) {
			logger.error(
				`Workspace symbols failed: ${error instanceof Error ? error.message : String(error)}`,
			);
			return [];
		}
	});
}

/**
 * Helper to get selection range (key only) or fall back to full range.
 * 
 * The selection range MUST be contained within the full range per LSP spec.
 * In YAML, getRange() returns the value's range, not including the key.
 * getKeyRange() returns just the key's range.
 * Since these don't overlap, we fall back to using the full range for selection.
 */
function getSelectionRange(
	cache: DocumentCache,
	cached: CachedDocument,
	path: (string | number)[],
	fullRange: Range,
): Range {
	const keyRange = cache.getKeyRange(cached, path);
	
	// Check if keyRange is within fullRange
	if (keyRange) {
		const keyStart = keyRange.start.line * 10000 + keyRange.start.character;
		const keyEnd = keyRange.end.line * 10000 + keyRange.end.character;
		const fullStart = fullRange.start.line * 10000 + fullRange.start.character;
		const fullEnd = fullRange.end.line * 10000 + fullRange.end.character;
		
		// Only use keyRange if it's contained within fullRange
		if (keyStart >= fullStart && keyEnd <= fullEnd) {
			return keyRange;
		}
	}
	
	// Fall back to using fullRange for selection
	// Create a minimal selection at the start of the range
	return {
		start: fullRange.start,
		end: {
			line: fullRange.start.line,
			character: fullRange.start.character + 1,
		},
	};
}

/**
 * Provide OpenAPI-specific document symbols (semantic outline).
 */
function provideOpenAPIDocumentSymbols(
	cached: CachedDocument,
	cache: DocumentCache,
	_ctx: TelescopeContext,
): DocumentSymbol[] {
	const symbols: DocumentSymbol[] = [];
	const ast = cached.parsedObject as Record<string, unknown>;

	// Add info section
	if (ast.info) {
		const infoRange = cache.getRange(cached, ["info"]);
		if (infoRange) {
			const info = ast.info as Record<string, unknown>;
			symbols.push({
				name: "info",
				detail: String(info.title ?? ""),
				kind: SymbolKind.Module,
				range: infoRange,
				selectionRange: getSelectionRange(cache, cached, ["info"], infoRange),
				children: [],
			});
		}
	}

	// Add paths
	const paths = ast.paths as Record<string, unknown> | undefined;
	if (paths) {
		const pathsRange = cache.getRange(cached, ["paths"]);
		if (pathsRange) {
			const pathChildren: DocumentSymbol[] = [];

			for (const [path, pathItem] of Object.entries(paths)) {
				if (!pathItem || typeof pathItem !== "object") continue;

				const pathRange = cache.getRange(cached, ["paths", path]);
				if (!pathRange) continue;

				const operationChildren: DocumentSymbol[] = [];
				const methods = [
					"get",
					"post",
					"put",
					"patch",
					"delete",
					"options",
					"head",
				];

				for (const method of methods) {
					const operation = (pathItem as Record<string, unknown>)[method] as
						| Record<string, unknown>
						| undefined;
					if (!operation) continue;

					const methodRange = cache.getRange(cached, ["paths", path, method]);
					if (!methodRange) continue;

					operationChildren.push({
						name: method.toUpperCase(),
						detail: String(
							operation.operationId ?? operation.summary ?? "",
						),
						kind: SymbolKind.Method,
						range: methodRange,
						selectionRange: getSelectionRange(cache, cached, ["paths", path, method], methodRange),
					});
				}

				pathChildren.push({
					name: path,
					kind: SymbolKind.Namespace,
					range: pathRange,
					selectionRange: getSelectionRange(cache, cached, ["paths", path], pathRange),
					children: operationChildren,
				});
			}

			symbols.push({
				name: "paths",
				kind: SymbolKind.Package,
				range: pathsRange,
				selectionRange: getSelectionRange(cache, cached, ["paths"], pathsRange),
				children: pathChildren,
			});
		}
	}

	// Add components
	const components = ast.components as Record<string, unknown> | undefined;
	if (components) {
		const componentsRange = cache.getRange(cached, ["components"]);
		if (componentsRange) {
			const componentChildren: DocumentSymbol[] = [];

			const sections = [
				{ key: "schemas", kind: SymbolKind.Class },
				{ key: "parameters", kind: SymbolKind.Property },
				{ key: "responses", kind: SymbolKind.Event },
				{ key: "requestBodies", kind: SymbolKind.Struct },
				{ key: "headers", kind: SymbolKind.Field },
				{ key: "securitySchemes", kind: SymbolKind.Key },
				{ key: "links", kind: SymbolKind.Interface },
				{ key: "callbacks", kind: SymbolKind.Function },
			] as const;

			for (const { key, kind } of sections) {
				const section = components[key] as Record<string, unknown> | undefined;
				if (!section) continue;

				const sectionRange = cache.getRange(cached, ["components", key]);
				if (!sectionRange) continue;

				const sectionChildren: DocumentSymbol[] = [];

				for (const name of Object.keys(section)) {
					const itemRange = cache.getRange(cached, ["components", key, name]);
					if (!itemRange) continue;

					sectionChildren.push({
						name,
						kind,
						range: itemRange,
						selectionRange: getSelectionRange(cache, cached, ["components", key, name], itemRange),
					});
				}

				componentChildren.push({
					name: key,
					kind: SymbolKind.Namespace,
					range: sectionRange,
					selectionRange: getSelectionRange(cache, cached, ["components", key], sectionRange),
					children: sectionChildren,
				});
			}

			symbols.push({
				name: "components",
				kind: SymbolKind.Package,
				range: componentsRange,
				selectionRange: getSelectionRange(cache, cached, ["components"], componentsRange),
				children: componentChildren,
			});
		}
	}

	// Add tags
	const tags = ast.tags as Array<{ name: string }> | undefined;
	if (tags && tags.length > 0) {
		const tagsRange = cache.getRange(cached, ["tags"]);
		if (tagsRange) {
			const tagChildren: DocumentSymbol[] = [];

			for (let i = 0; i < tags.length; i++) {
				const tag = tags[i];
				const tagRange = cache.getRange(cached, ["tags", i]);
				if (!tagRange) continue;

				tagChildren.push({
					name: tag.name,
					kind: SymbolKind.EnumMember,
					range: tagRange,
					selectionRange: tagRange, // Tags don't have keys, use full range
				});
			}

			symbols.push({
				name: "tags",
				kind: SymbolKind.Enum,
				range: tagsRange,
				selectionRange: getSelectionRange(cache, cached, ["tags"], tagsRange),
				children: tagChildren,
			});
		}
	}

	return symbols;
}

/**
 * Provide workspace symbols matching a query.
 *
 * @param query - Search query to filter symbols
 * @param ctx - Telescope context for accessing known OpenAPI files
 * @param provider - Document provider for fetching and parsing documents
 * @param reportPartial - Optional callback to report partial results incrementally
 */
async function provideWorkspaceSymbols(
	query: string,
	ctx: TelescopeContext,
	provider: DocumentProvider,
	reportPartial?: (symbols: WorkspaceSymbol[]) => void,
): Promise<WorkspaceSymbol[]> {
	const allSymbols: WorkspaceSymbol[] = [];
	const lowerQuery = query.toLowerCase();

	// Batch size for partial result reporting (symbols per batch)
	const PARTIAL_BATCH_SIZE = 50;
	let pendingSymbols: WorkspaceSymbol[] = [];

	const flushPartial = () => {
		if (reportPartial && pendingSymbols.length > 0) {
			reportPartial(pendingSymbols);
			pendingSymbols = [];
		}
	};

	const addSymbol = (symbol: WorkspaceSymbol) => {
		allSymbols.push(symbol);
		if (reportPartial) {
			pendingSymbols.push(symbol);
			if (pendingSymbols.length >= PARTIAL_BATCH_SIZE) {
				flushPartial();
			}
		}
	};

	for (const uri of ctx.getKnownOpenAPIFiles()) {
		const doc = await provider.get(uri);
		if (!doc) continue;

		const ast = getAstForProvided(doc);
		if (!ast || typeof ast !== "object") continue;

		// Search operations
		const paths = (ast as any).paths as Record<string, unknown> | undefined;
		if (paths) {
			const methods = [
				"get",
				"post",
				"put",
				"patch",
				"delete",
				"options",
				"head",
			];

			for (const [path, pathItem] of Object.entries(paths)) {
				if (!pathItem || typeof pathItem !== "object") continue;

				for (const method of methods) {
					const operation = (pathItem as Record<string, unknown>)[method] as
						| Record<string, unknown>
						| undefined;
					if (!operation) continue;

					const opId = operation.operationId
						? String(operation.operationId)
						: `${method.toUpperCase()} ${path}`;

					if (opId.toLowerCase().includes(lowerQuery)) {
						const pointer = joinPointer(["paths", path, method]);
						const range = provider.pointerToRange(doc, pointer);
						if (range) {
							addSymbol({
								name: opId,
								kind: SymbolKind.Method,
								location: { uri, range },
								containerName: path,
							});
						}
					}
				}
			}
		}

		// Search components
		const components = (ast as any).components as Record<string, unknown> | undefined;
		if (components) {
			const sections = [
				{ key: "schemas", kind: SymbolKind.Class },
				{ key: "parameters", kind: SymbolKind.Property },
				{ key: "responses", kind: SymbolKind.Event },
			] as const;

			for (const { key, kind } of sections) {
				const section = components[key] as Record<string, unknown> | undefined;
				if (!section) continue;

				for (const name of Object.keys(section)) {
					if (name.toLowerCase().includes(lowerQuery)) {
						const pointer = joinPointer(["components", key, name]);
						const range = provider.pointerToRange(doc, pointer);
						if (range) {
							addSymbol({
								name,
								kind,
								location: { uri, range },
								containerName: `components/${key}`,
							});
						}
					}
				}
			}
		}

		// After processing each file, flush any pending partial results
		// This ensures responsive updates even for large files
		flushPartial();
	}

	// Final flush for any remaining symbols
	flushPartial();

	return allSymbols;
}

function getAstForProvided(doc: ProvidedDocument): unknown {
	return doc.kind === "open" ? doc.cached.parsedObject : doc.parsed.ast;
}

File: server/lsp/symbols.go
package lsp

import (
	"fmt"
	"strings"

	"github.com/LukasParke/gossip"
	"github.com/LukasParke/gossip/protocol"
	"github.com/sailpoint-oss/telescope/server/lsp/adapt"
	"github.com/sailpoint-oss/telescope/server/openapi"
)

// clampSelection ensures selectionRange is contained within fullRange.
// If it isn't (or is zero-valued) fullRange is returned as the selection.
func clampSelection(fullRange, selRange protocol.Range) protocol.Range {
	if rangeContains(fullRange, selRange) {
		return selRange
	}
	return fullRange
}

func rangeContains(outer, inner protocol.Range) bool {
	if positionBefore(inner.Start, outer.Start) {
		return false
	}
	if positionBefore(outer.End, inner.End) {
		return false
	}
	return true
}

func positionBefore(a, b protocol.Position) bool {
	return a.Line < b.Line || (a.Line == b.Line && a.Character < b.Character)
}

// NewSymbolHandler provides document symbols for the OpenAPI structure.
func NewSymbolHandler(cache *openapi.IndexCache, _ *GraphBridge) gossip.DocumentSymbolHandler {
	return func(ctx *gossip.Context, params *protocol.DocumentSymbolParams) ([]protocol.DocumentSymbol, error) {
		idx := cache.Get(params.TextDocument.URI)
		if idx == nil || !idx.IsOpenAPI() {
			return nil, nil
		}

		var symbols []protocol.DocumentSymbol

		// Info
		if idx.Document.Info != nil {
			info := idx.Document.Info
			title := info.Title
			if title == "" {
				title = "(untitled)"
			}
			detail := ""
			if info.Version != "" {
				detail = "v" + info.Version
			}
			symbols = append(symbols, protocol.DocumentSymbol{
				Name:           title,
				Detail:         detail,
				Kind:           protocol.SymbolPackage,
				Range:          adapt.RangeToProtocol(info.Loc.Range),
				SelectionRange: clampSelection(adapt.RangeToProtocol(info.Loc.Range), adapt.RangeToProtocol(info.TitleLoc.Range)),
			})
		}

		// Paths
		if len(idx.Document.Paths) > 0 {
			var pathChildren []protocol.DocumentSymbol
			for path, item := range idx.Document.Paths {
				pathSymbol := protocol.DocumentSymbol{
					Name:           path,
					Kind:           protocol.SymbolModule,
					Range:          adapt.RangeToProtocol(item.Loc.Range),
					SelectionRange: clampSelection(adapt.RangeToProtocol(item.Loc.Range), adapt.RangeToProtocol(item.PathLoc.Range)),
				}
				for _, mo := range item.Operations() {
					name := strings.ToUpper(mo.Method)
					if mo.Operation.OperationID != "" {
						name = fmt.Sprintf("%s (%s)", name, mo.Operation.OperationID)
					}
					pathSymbol.Children = append(pathSymbol.Children, protocol.DocumentSymbol{
						Name:           name,
						Kind:           protocol.SymbolMethod,
						Range:          adapt.RangeToProtocol(mo.Operation.Loc.Range),
						SelectionRange: adapt.RangeToProtocol(mo.Operation.Loc.Range),
					})
				}
				pathChildren = append(pathChildren, pathSymbol)
			}
			symbols = append(symbols, protocol.DocumentSymbol{
				Name:           "paths",
				Kind:           protocol.SymbolNamespace,
				Range:          adapt.RangeToProtocol(idx.Document.Loc.Range),
				SelectionRange: adapt.RangeToProtocol(idx.Document.Loc.Range),
				Children:       pathChildren,
			})
		}

		if idx.Document.Components == nil {
			return appendTagSymbols(symbols, idx), nil
		}
		comp := idx.Document.Components

		// Schemas
		if len(comp.Schemas) > 0 {
			var children []protocol.DocumentSymbol
			for name, schema := range comp.Schemas {
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         schema.Type,
					Kind:           protocol.SymbolClass,
					Range:          adapt.RangeToProtocol(schema.Loc.Range),
					SelectionRange: clampSelection(adapt.RangeToProtocol(schema.Loc.Range), adapt.RangeToProtocol(schema.NameLoc.Range)),
				})
			}
			symbols = append(symbols, componentGroupSymbol("schemas", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Parameters
		if len(comp.Parameters) > 0 {
			var children []protocol.DocumentSymbol
			for name, param := range comp.Parameters {
				detail := param.In
				if param.Required {
					detail += " (required)"
				}
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         detail,
					Kind:           protocol.SymbolVariable,
					Range:          adapt.RangeToProtocol(param.Loc.Range),
					SelectionRange: clampSelection(adapt.RangeToProtocol(param.Loc.Range), adapt.RangeToProtocol(param.NameLoc.Range)),
				})
			}
			symbols = append(symbols, componentGroupSymbol("parameters", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Responses
		if len(comp.Responses) > 0 {
			var children []protocol.DocumentSymbol
			for name, resp := range comp.Responses {
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         truncate(resp.Description.Text, 60),
					Kind:           protocol.SymbolField,
					Range:          adapt.RangeToProtocol(resp.Loc.Range),
					SelectionRange: adapt.RangeToProtocol(resp.Loc.Range),
				})
			}
			symbols = append(symbols, componentGroupSymbol("responses", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Request Bodies
		if len(comp.RequestBodies) > 0 {
			var children []protocol.DocumentSymbol
			for name, rb := range comp.RequestBodies {
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         truncate(rb.Description.Text, 60),
					Kind:           protocol.SymbolStruct,
					Range:          adapt.RangeToProtocol(rb.Loc.Range),
					SelectionRange: adapt.RangeToProtocol(rb.Loc.Range),
				})
			}
			symbols = append(symbols, componentGroupSymbol("requestBodies", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Headers
		if len(comp.Headers) > 0 {
			var children []protocol.DocumentSymbol
			for name, hdr := range comp.Headers {
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         truncate(hdr.Description.Text, 60),
					Kind:           protocol.SymbolField,
					Range:          adapt.RangeToProtocol(hdr.Loc.Range),
					SelectionRange: adapt.RangeToProtocol(hdr.Loc.Range),
				})
			}
			symbols = append(symbols, componentGroupSymbol("headers", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Security Schemes
		if len(comp.SecuritySchemes) > 0 {
			var children []protocol.DocumentSymbol
			for name, ss := range comp.SecuritySchemes {
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         ss.Type,
					Kind:           protocol.SymbolProperty,
					Range:          adapt.RangeToProtocol(ss.Loc.Range),
					SelectionRange: adapt.RangeToProtocol(ss.Loc.Range),
				})
			}
			symbols = append(symbols, componentGroupSymbol("securitySchemes", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Links
		if len(comp.Links) > 0 {
			var children []protocol.DocumentSymbol
			for name, link := range comp.Links {
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Detail:         truncate(link.Description.Text, 60),
					Kind:           protocol.SymbolProperty,
					Range:          adapt.RangeToProtocol(link.Loc.Range),
					SelectionRange: adapt.RangeToProtocol(link.Loc.Range),
				})
			}
			symbols = append(symbols, componentGroupSymbol("links", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		// Callbacks
		if len(comp.Callbacks) > 0 {
			var children []protocol.DocumentSymbol
			for name, cb := range comp.Callbacks {
				r := callbackRange(*cb)
				children = append(children, protocol.DocumentSymbol{
					Name:           name,
					Kind:           protocol.SymbolFunction,
					Range:          r,
					SelectionRange: r,
				})
			}
			symbols = append(symbols, componentGroupSymbol("callbacks", adapt.RangeToProtocol(comp.Loc.Range), children))
		}

		return appendTagSymbols(symbols, idx), nil
	}
}

func appendTagSymbols(symbols []protocol.DocumentSymbol, idx *openapi.Index) []protocol.DocumentSymbol {
	if len(idx.Tags) == 0 {
		return symbols
	}
	var children []protocol.DocumentSymbol
	for _, tag := range idx.Document.Tags {
		children = append(children, protocol.DocumentSymbol{
			Name:           tag.Name,
			Detail:         truncate(tag.Description.Text, 60),
			Kind:           protocol.SymbolString,
			Range:          adapt.RangeToProtocol(tag.Loc.Range),
			SelectionRange: clampSelection(adapt.RangeToProtocol(tag.Loc.Range), adapt.RangeToProtocol(tag.NameLoc.Range)),
		})
	}
	symbols = append(symbols, protocol.DocumentSymbol{
		Name:           "tags",
		Kind:           protocol.SymbolNamespace,
		Range:          adapt.RangeToProtocol(idx.Document.Loc.Range),
		SelectionRange: adapt.RangeToProtocol(idx.Document.Loc.Range),
		Children:       children,
	})
	return symbols
}

func componentGroupSymbol(name string, parentRange protocol.Range, children []protocol.DocumentSymbol) protocol.DocumentSymbol {
	return protocol.DocumentSymbol{
		Name:           name,
		Kind:           protocol.SymbolNamespace,
		Range:          parentRange,
		SelectionRange: parentRange,
		Children:       children,
	}
}

func callbackRange(cb openapi.Callback) protocol.Range {
	for _, item := range cb {
		if item != nil {
			return adapt.RangeToProtocol(item.Loc.Range)
		}
	}
	return protocol.Range{}
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max-3] + "..."
}
